### PR TITLE
Reflection Based Testing + Removed LinkedList

### PR DESCRIPTION
**Pull request by MosheWolberg**
Implemented a new testing methodology that uses reflection based off discussions with Aaron. 
Changed the way method handles are added to Type definitions (no more LinkedList) based off discussions with Jan which resolved bugs in edge cases.

### CODE REVIEW

File: src/Experiment/src/AssemblyBuilder.cs [hunk 1]
```diff
@@ -10,22 +9,22 @@
...
 
-    public class AssemblyBuilder: System.Reflection.Assembly
+    public class AssemblyBuilder : System.Reflection.Assembly
     {
-        private bool _previouslySaved=false;
-        private AssemblyName _assemblyName;
-        private MetadataBuilder _metadata;
+        private bool _previouslySaved = false;
+        private AssemblyName? _assemblyName;
+        private MetadataBuilder metadata;
```
Comment (jkotas) [inline src/Experiment/src/AssemblyBuilder.cs]:
```suggestion
```

File: src/Experiment/src/AssemblyBuilder.cs [hunk 2]
```diff
@@ -10,22 +9,22 @@
...
         private ModuleBuilder? _module;
 
-        private AssemblyBuilder(AssemblyName name) 
+        private AssemblyBuilder(AssemblyName name)
         {
             _assemblyName = name;
-            _metadata = new MetadataBuilder();
+            metadata = new MetadataBuilder();
```
Comment (jkotas) [inline src/Experiment/src/AssemblyBuilder.cs]:
```suggestion
```

File: src/Experiment/src/AssemblyBuilder.cs [hunk 3]
```diff
@@ -39,27 +38,27 @@ public void Save(string assemblyFileName)
             {
                 throw new ArgumentNullException(nameof(_assemblyName));
             }
-            
-            if(_module == null)
+
+            if (_module == null)
             {
                 throw new InvalidOperationException("Assembly needs at least one module defined");
             }
```
Comment (jkotas) [inline src/Experiment/src/AssemblyBuilder.cs:44]:
```suggestion
            var metadata = new MetadataBuilder();

```

File: src/Experiment/src/AssemblyBuilder.cs [hunk 4]
```diff
@@ -39,27 +38,27 @@ public void Save(string assemblyFileName)
...
 
             //Add assembly metadata
-            _metadata.AddAssembly(//Metadata is added for the new assembly - Current design - metadata generated only when Save method is called.
-               _metadata.GetOrAddString(value: _assemblyName.Name),
+            metadata.AddAssembly(//Metadata is added for the new assembly - Current design - metdata generated only when Save method is called.
```
Comment (jkotas) [inline src/Experiment/src/AssemblyBuilder.cs]:
```suggestion
            metadata.AddAssembly(//Metadata is added for the new assembly - Current design - metadata generated only when Save method is called.
```

File: src/Experiment/src/AssemblyBuilder.cs [hunk 5]
```diff
@@ -39,27 +38,27 @@ public void Save(string assemblyFileName)
...
+               metadata.GetOrAddString(value: _assemblyName.Name),
                version: _assemblyName.Version ?? new Version(0, 0, 0, 0),
-               culture: (_assemblyName.CultureName==null) ?  default : _metadata.GetOrAddString(value: _assemblyName.CultureName),
-               publicKey: (_assemblyName.GetPublicKey() is byte[] publicKey) ? _metadata.GetOrAddBlob(value: publicKey) : default,
-               flags: (AssemblyFlags) _assemblyName.Flags,
-               hashAlgorithm: AssemblyHashAlgorithm.None);//AssemblyName.HashAlgorithm is obsolete so default value used.
+               culture: (_assemblyName.CultureName == null) ? default : metadata.GetOrAddString(value: _assemblyName.CultureName),
+               publicKey: (_assemblyName.GetPublicKey() is byte[] publicKey) ? metadata.GetOrAddBlob(value: publicKey) : default,
+               flags: (AssemblyFlags)_assemblyName.Flags,
+               hashAlgorithm: AssemblyHashAlgorithm.None);//It seems AssemblyName.HashAlgorithm is obslete so default value used.
```
Comment (jkotas) [inline src/Experiment/src/AssemblyBuilder.cs]:
```suggestion
               hashAlgorithm: AssemblyHashAlgorithm.None);//It seems AssemblyName.HashAlgorithm is obsolete so default value used.
```

File: src/Experiment/src/AssemblyBuilder.cs [hunk 6]
```diff
@@ -98,29 +97,29 @@ public System.Reflection.Emit.Experimental.ModuleBuilder DefineDynamicModule(str
...
-            if(name == null)
+            if (name == null)
             {
                 throw new ArgumentNullException(nameof(name));
             }
 
             if (name.Length == 0)
             {
-                throw new ArgumentException(nameof(name));
+                throw new ArgumentException($"{nameof(name)} cannot have zero charachters.");
```
Comment (jkotas) [inline src/Experiment/src/AssemblyBuilder.cs]:
```suggestion
                throw new ArgumentException($"{nameof(name)} cannot have zero characters.");
```

File: src/Experiment/tests/AssemblyLoadTools.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
+﻿using System;
```
Comment (jkotas) [inline src/Experiment/tests/AssemblyLoadTools.cs]:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

﻿using System;
```

File: src/Experiment/src/AssemblyBuilder.cs [hunk 8]
```diff
@@ -10,22 +9,20 @@
...
 {
 
-    public class AssemblyBuilder: System.Reflection.Assembly
+    public class AssemblyBuilder : System.Reflection.Assembly
     {
-        private bool _previouslySaved=false;
-        private AssemblyName _assemblyName;
-        private MetadataBuilder _metadata;
+        private bool _previouslySaved = false;
+        private AssemblyName? _assemblyName;
```
Comment (joperezr) [inline src/Experiment/src/AssemblyBuilder.cs:15]: Why do we need to add nullable annotation here when the only constructor doesn't take a nullable `AssemblyName`?

**Comment by ericstj** [comment]
@MosheWolberg can you have a look at @joperezr's feedback and move this along?

**Pull request merged**